src/django_components/component_media.py

```diff
@@ -0,0 +1,378 @@
+    This class allows to override which class will be instantiated with `media_class` attribute:
+
+    ```py
+    class MyMedia(Media):
+        def render_js(self):
+            ...
+
+    class MyComponent(component.Component):
+        media_class = MyMedia
+        def get_context_data(self):
+            assert isinstance(self.media, MyMedia)
+    ```
+    """
+
+    def __new__(mcs, name: str, bases: Tuple[Type, ...], attrs: Dict[str, Any]) -> Type:
+        if "Media" in attrs:
+            media_data: ComponentMediaInput = attrs["Media"]
+            # Normalize the various forms of Media inputs we allow
+            _normalize_media(media_data)
+            # Given a predictable structure of Media class, get all the various JS/CSS paths
+            # that user has defined, and normalize them too.
+            #
+            # Because we can accept:
+            # str, bytes, PathLike, SafeData (AKA Django's "path as object") or a callable
+            #
+            # And we want to convert that to:
+            # str and SafeData
+            _map_media_filepaths(media_data, _normalize_media_filepath)
+
+        # Once the inputs are normalized, attempt to resolve the JS/CSS filepaths
+        # as relative to the directory where the component class is defined.
+        _resolve_component_relative_files(attrs)
+
+        # Since we're inheriting from `MediaDefiningClass`, it should take the inputs
+        # from `cls.Media`, and set the `cls.media` to an instance of Django's `Media` class
+        cls = super().__new__(mcs, name, bases, attrs)
+
+        # Lastly, if the class defines `media_class` attribute, transform `cls.media`
+        # to the instance of `media_class`.
+        _monkeypatch_media_property(cls)
+
+        return cls
+
+
+# Allow users to provide custom subclasses of Media via `media_class`.
+# `MediaDefiningClass` defines `media` as a getter (defined in django.forms.widgets.media_property).
+# So we reused that and convert it to user-defined Media class
+def _monkeypatch_media_property(comp_cls: Type["Component"]) -> None:
+    if not hasattr(comp_cls, "media_class"):
+        return
+
+    media_prop: property = comp_cls.media
+    media_getter = media_prop.fget
+
+    def media_wrapper(self: "Component") -> Any:
+        if not media_getter:
+            return None
+        media: Media = media_getter(self)
+        return self.media_class(js=media._js, css=media._css)
+
+    comp_cls.media = property(media_wrapper)
+
+
+def _normalize_media(media: ComponentMediaInput) -> None:
+    if hasattr(media, "css") and media.css:
+        # Allow: class Media: css = "style.css"
+        if _is_media_filepath(media.css):
+            media.css = [media.css]  # type: ignore[list-item]
+
+        # Allow: class Media: css = ["style.css"]
+        if isinstance(media.css, (list, tuple)):
+            media.css = {"all": media.css}
```

Comment:
Are we changing the media type from "None" to "all" here? media="all" is implicit in CSS so I don't think we should add that to everyone's CSS when it's not needed.

Reply:
I didn't implement it, so I might be missing something, but I'm not aware of the "None" option, was that maybe something in older Django versions? Because in v5, when you [go to the docs](https://docs.djangoproject.com/en/5.0/topics/forms/media/#css) and see the examples there, it looks like `"all"` is the default / catch-all option.

Reply:
Btw @EmilStenstrom I'll merge this MR so we can close the original issue. But if there's a way to not generate the `media="all"`, e.g. with the "None" option, we can open a new issue for that or I can just include it in on of my other MRs

Reply:
I was thinking that `css = "style.css"` (our version) would generate a link tag without media set. But let's keep it the way to wrote it.